docs/setup/upgrade/resolving-migration-failures.asciidoc

```diff
@@ -106,7 +105,21 @@ If you fail to remedy this, your upgrade to 8.0+ will fail with a message like:
 
 [source,sh]
 --------------------------------------------
-Unable to complete saved object migrations for the [.kibana] index: Migration failed because documents were found for unknown saved object types. To proceed with the migration, please delete these documents from the ".kibana_7.17.0_001" index.
+Unable to complete saved object migrations for the [.kibana] index: Migration failed because some documents were found which use unknown saved object types:
+- "firstDocId" (type "someType")
+- "secondtDocId" (type "someType")
+- "thirdDocId" (type "someOtherType")
+- ... # non-exhaustive list of unknown object ids and their types>
```

Comment:
`non-exhaustive` or `exhaustive`? We're listing all the docs in the actual log, aren't we?

Reply:
we're listing the 1000 first results, but there could be more.

Reply:
Yeah right now if there are more than 1000 results we might not detect all types so migrations might consistently fail if unknown types weren't detected and therefore not ignored.

We also end up discarding documents which we didn't warn the user about.

I wonder if we a top_hits aggregation might not be better like:
```
GET /.kibana/_search
{
  "size": 0,
  "aggs": {
    "district": {
      "terms": {
        "field": "type",
        "size": 100
      },
      "aggs": {
        "docs": {
          "top_hits": {
            "size": 100,
            "_source": {
              "includes": ["updated_at"]
            }
          }
        }
      }
    }
  }
}
```

Which will produce output like:
```json
"aggregations" : {
    "district" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [
        {
          "key" : "epm-packages-assets",
          "doc_count" : 463,
          "docs" : {
            "hits" : {
              "total" : {
                "value" : 463,
                "relation" : "eq"
              },
              "max_score" : 1.0,
              "hits" : [
                {
                  "_index" : ".kibana_8.0.1_001",
                  "_id" : "epm-packages-assets:3a880207-215b-5bef-8449-5a2411652118",
                  "_score" : 1.0,
                  "_source" : {
                    "updated_at" : "2022-03-07T14:25:49.908Z"
                  }
                },
                {
                  "_index" : ".kibana_8.0.1_001",
                  "_id" : "epm-packages-assets:8b983072-c517-5e78-9b1c-d14f441028b6",
                  "_score" : 1.0,
                  "_source" : {
                    "updated_at" : "2022-03-07T14:25:49.908Z"
                  }
                },
```


There's two limitations, the bucket aggregation size means if there are more than 100 unknown saved object types some might go unnoticed. It's hard to imagine this happening in the real world.

top_hits cannot show more than 100 hits per bucket so we can't show all _ids. But probably what's more important for users and for our exclusion filter is to have a complete list of types. Because unknown types are linked to disabled plugins users usually want all or none of the documents of a certain type.

The other alternative is to use the `Actions.readWithPit` to iterate through all documents, but it should be faster to let Elasticsearch do the work for us.

Reply:
I thought having updated_at might be useful so added `"includes": ["updated_at"]`, but maybe the logs just become way too verbose and we should just use `"excludes": ["*"]` so that we only get the _id in the aggregation.

Reply:
Thanks! I'll add your aggregation proposal to the query and transform results / log accordingly.

Reply:
@rudolf @pgayvallet  I was looking at the current transform logic, and it contemplates a case where `type` is not known:

https://github.com/elastic/kibana/blob/main/src/core/server/saved_objects/deprecations/unknown_object_types.ts#L86

I don't know whether this can really happen, and if it can happen, I wonder how the aggregation proposed by rudolf would behave in that case.

Reply:
I have split the "unknown types" detection in 2 separate queries:

- One to detect those documents that don't define a `type` attribute.
- Another one to detect those documents that define a type that is not part of the registered types.

On the same lines, I have improved the `excludeFromUpgrade` query: now it also excludes those documents that do NOT have a type.

Reply:
Good you thought about the missing types, it's an edge case but we have had SDH's where the type field was missing.

It's possible to do a bucket aggregation over keys that don't exist by specifying the `missing` parameter like:
```
GET /test/_search
{
  "size": 0,
  "aggs": {
    "types": {
      "terms": {
        "field": "type",
        "size": 1000,
        "missing": "__UNKNOWN__"
      },
      "aggs": {
        "docs": {
          "top_hits": {
            "size": 100,
            "_source": {
              "includes": ["updated_at"]
            }
          }
        }
      }
    }
  }
}
```

That means that if anyone ever registers a saved object type `__UNKNOWN__` it will fail but this would be evident in development and rather unlikely.

Reply:
Cool! that will allow retrieving all unknown types (unregistered + undefined) in a single query.
Simplified in my last commits.